backend/stakepoold/rpc/rpcserver/client.go

```diff
@@ -0,0 +1,168 @@
+package rpcserver
+
+import (
+	"sync"
+	"time"
+
+	"github.com/decred/dcrd/rpcclient/v3"
+)
+
+const (
+	// connectionRetryInterval is the amount of time to wait in between
+	// retries when automatically reconnecting to an RPC server.
+	connectionRetryInterval = time.Second * 5
+	// disconnectCheckInterval is the amount of time to wait between
+	// checks for a disconnection.
+	disconnectCheckInterval = time.Second * 10
+)
+
+// Client holds the information related to an rpcclient and handles access to
+// that client through a mutex.
+//
+// It should be noted that this is a temporary fix to the problem that rpcclient
+// does not return an error when autoreconnect is turned on but the client is
+// disconnected. The permanent solution is to change the behaviour of rpccleint.
+// TODO: Remove this file.
+type Client struct {
+	// client is protected by a mutex that must be held for reads/writes.
+	client *rpcclient.Client
+	mux    sync.RWMutex
+
+	cfg          *rpcclient.ConnConfig
+	ntfnHandlers *rpcclient.NotificationHandlers
+	wg           sync.WaitGroup
+	stop         chan struct{}
+
+	connected    chan struct{}
+	connectedMux sync.Mutex
+}
+
+// Connected returns a receiving copy of the current connected channel. If
+// disconnected and the channel is not yet blocking, creates a new channel that
+// will be closed on a successful reconnect.
+func (c *Client) Connected() <-chan struct{} {
+	c.connectedMux.Lock()
+	defer c.connectedMux.Unlock()
+	if c.RPCClient().Disconnected() {
+		// Start blocking on connected chan if not already.
+		select {
+		case <-c.connected:
+			c.connected = make(chan struct{})
+		default:
+		}
+	}
+	return c.connected
+}
+
+// RPCClient allows access to the underlying rpcclient by providing a copy of
+// its address.
+func (c *Client) RPCClient() *rpcclient.Client {
+	c.mux.RLock()
+	defer c.mux.RUnlock()
+	return c.client
+}
+
+// New creates a new Client and starts the automatic reconnection handler.
+// Returns an error if unable to construct a new rpcclient.
+func New(cfg *rpcclient.ConnConfig, ntfnHandlers *rpcclient.NotificationHandlers) (*Client, error) {
+	client, err := rpcclient.New(cfg, ntfnHandlers)
+	if err != nil {
+		return nil, err
+	}
+	c := &Client{
+		client:       client,
+		cfg:          cfg,
+		ntfnHandlers: ntfnHandlers,
+		stop:         make(chan struct{}),
+		connected:    make(chan struct{}),
+	}
+	// A closed connected channel indcates successfully connected.
+	close(c.connected)
+	go func(c *Client) {
+		c.wg.Add(1)
+		c.autoReconnect()
+	}(c)
```

Comment:
instead of this closure as a goroutine, how about

```go
c.wg.Add(1)
go c.autoReconnect()
```

But even if you do have a func here, it's not necessary to pass `c` as an argument since it will be captured by the closure.